util/patch/patch_test.go

```diff
@@ -997,3 +999,1683 @@ func TestNewHelperNil(t *testing.T) {
+		})
+
+		t.Run("should recover if there is a resolvable conflict", func(t *testing.T) {
+			g := NewWithT(t)
+
+			obj := obj.DeepCopy()
+
+			t.Log("Creating the object")
+			g.Expect(env.Create(ctx, obj)).To(Succeed())
+			defer func() {
+				g.Expect(env.Delete(ctx, obj)).To(Succeed())
+			}()
+			key := client.ObjectKey{Name: obj.Name, Namespace: obj.Namespace}
+
+			t.Log("Checking that the object has been created")
+			g.Eventually(func() error {
+				obj := obj.DeepCopy()
+				return env.Get(ctx, key, obj)
+			}).Should(Succeed())
+
+			objCopy := obj.DeepCopy()
+
+			t.Log("Marking condition and back compatibility condition Test=False")
+			conditions.MarkFalse(objCopy, clusterv1.ConditionType("Test"), "reason", clusterv1.ConditionSeverityInfo, "message")
+			err := v1beta2conditions.Set(objCopy, metav1.Condition{Type: "Test", Status: metav1.ConditionFalse, Reason: "reason", Message: "message"})
+			g.Expect(err).ToNot(HaveOccurred())
+			g.Expect(env.Status().Update(ctx, objCopy)).To(Succeed())
+
+			t.Log("Validating that the local object's resource version is behind")
+			g.Expect(obj.ResourceVersion).NotTo(Equal(objCopy.ResourceVersion))
+
+			t.Log("Creating a new patch helper")
+			patcher, err := NewHelper(obj, env)
+			g.Expect(err).ToNot(HaveOccurred())
+
+			t.Log("Marking condition and back compatibility condition Ready=True")
+			conditions.MarkTrue(obj, clusterv1.ReadyCondition)
+			err = v1beta2conditions.Set(obj, metav1.Condition{Type: "Ready", Status: metav1.ConditionTrue, Reason: "AllGood"})
+			g.Expect(err).ToNot(HaveOccurred())
+
+			t.Log("Patching the object")
+			g.Expect(patcher.Patch(ctx, obj)).To(Succeed())
+
+			t.Log("Validating the object has been updated")
+			g.Eventually(func() bool {
+				objAfter := obj.DeepCopy()
+				if err := env.Get(ctx, key, objAfter); err != nil {
+					return false
+				}
+
+				testBackCompatibilityCopy := conditions.Get(objCopy, "Test")
+				testBackCompatibilityAfter := conditions.Get(objAfter, "Test")
+				ok, err := conditions.MatchCondition(*testBackCompatibilityCopy).Match(*testBackCompatibilityAfter)
+				if err != nil || !ok {
+					return false
+				}
+
+				readyBackCompatibilityBefore := conditions.Get(obj, clusterv1.ReadyCondition)
+				readyBackCompatibilityAfter := conditions.Get(objAfter, clusterv1.ReadyCondition)
+				ok, err = conditions.MatchCondition(*readyBackCompatibilityBefore).Match(*readyBackCompatibilityAfter)
+				if err != nil || !ok {
+					return false
+				}
+
+				testConditionCopy, err := v1beta2conditions.Get(objCopy, "Test")
+				if err != nil {
+					return false
+				}
+				testConditionAfter, err := v1beta2conditions.Get(objAfter, "Test")
+				if err != nil {
+					return false
+				}
+				ok, err = v1beta2conditions.MatchCondition(*testConditionCopy, v1beta2conditions.IgnoreLastTransitionTime(true)).Match(*testConditionAfter)
```

Comment:
Why do we need Ignore here? We didn't need it for the corresponding phase 1 test

(I think we should audit all the IgnoreLastTransitionTime if they are correct in this file)

Reply:
Fixed (TLDR, we were not setting LastTransitionTime in the condition, so time.Now was applied introducing a diff between got and expected